ENH: reusing existing components, reorganizing, volume extent computation fix

che85 opened the pull request:
fixes #275

fedorov committed:
WIP: refactoring of the multiframe conversion implementation

This is an intermediate commit that aims to improve organization of the conversion
functionality, motivated by the various standing issues (e.g., #217, #192 and #150)
that are difficult to address with the current level of disarray in the code.

Some of the ideas are the following:

 - introduce basic hierarchy of classes to allow reuse of the common
   functionality between PM and SEG converters (MultiFrameConverter parent for
   ParametricMapConverter and SegmentationImageConverter)
 - do not use static functions, instead make better use of converter classes
 - factor out containers to keep information about ImageVolume and DICOMFrame
   to simplify conversion and book keeping
 - improve consistency of naming and code style

My plan to proceed with the refactoring is to implement "shadow" functions in the
updated classes while keeping the original functions initially, and then phase out
the original functions.

The current commit is just an intermediate milestone to open this for feedback
and keep development history at least somewhat incremental.

fedorov committed:
WIP: refactoring of the multiframe conversion implementation

This is an intermediate commit that aims to improve organization of the conversion
functionality, motivated by the various standing issues (e.g., #217, #192 and #150)
that are difficult to address with the current level of disarray in the code.

Some of the ideas are the following:

 - introduce basic hierarchy of classes to allow reuse of the common
   functionality between PM and SEG converters (MultiFrameConverter parent for
   ParametricMapConverter and SegmentationImageConverter)
 - do not use static functions, instead make better use of converter classes
 - factor out containers to keep information about ImageVolume and DICOMFrame
   to simplify conversion and book keeping
 - improve consistency of naming and code style

My plan to proceed with the refactoring is to implement "shadow" functions in the
updated classes while keeping the original functions initially, and then phase out
the original functions.

The current commit is just an intermediate milestone to open this for feedback
and keep development history at least somewhat incremental.

fedorov committed:
BUG: initialize modality from the source dataset

fedorov committed:
BUG: initialize modality from the source dataset

fedorov committed:
ENH: a different shot at refactoring

fedorov committed:
ENH: adding functionality - temp commit

fedorov committed:
ENH: more updates

fedorov committed:
ENH: checkpointing updates to the branch

fedorov committed:
BUG: fix API typo fixed in dcmtk

fedorov committed:
BUG: add docker_entry removed by accident in ff52820

fedorov committed:
BUG: return value

fedorov committed:
ENH: checkpoint commit

fedorov committed:
checkpoint

fedorov committed:
ENH: fix compile errors, initialize common instance module

fedorov committed:
ENH: fixing issues identified by tests

fedorov committed:
circle_ci another dummy

fedorov committed:
ENH: fix more issues

che85 committed:
BUG: fixing wrong initialization of DPMParametricMapIOD

che85 committed:
ENH: adding refactored code for paramap from DCM to itk

che85 committed:
ENH: adding refactored code for paramap from DCM to itk

che85 committed:
ENH: added region and origin to ImageVolumeGeometry getITKRepresentation

che85 committed:
ENH: initialize meta information from DCM into internal data structure

che85 committed:
ENH: complete initialization of meta information from DCM

* added Helper class static methods

che85 committed:
ENH: adaption to changes in https://github.com/QIICR/dcmqi/pull/271

che85 committed:
ENH: added first step for segmentation dcm to itk

che85 committed:
ENH: reading segmentAttributes into metadata

* added replacement method to SegmentationImageConverter

che85 committed:
ENH: ctests for dcm segmentation 2 itk succeed

* TODO: refactor even more and generalize

fedorov committed:
ENH: added initialization of derivation image FG

che85 committed:
ENH: extraction of more methods

fedorov committed:
ENH: added initialization of the frame content

fedorov committed:
ENH: fix initialization of the composite context

itkimage2paramap tests are passing!

che85 committed:
STYLE: and uncommented code

fedorov committed:
ENH: add RWVM FG

crash in addForAllFrames()!...

fedorov committed:
ENH: fix RWVM FG segfault

Also:
* fix the typo in JSON lookup
* replace the temporary codes with the standard ones: CP-1665 became standard
in 2017a

fedorov committed:
ENH: replace temporary codes with standard ones

CP-1665 became standard in 2017a

fedorov committed:
ENH: fix initialization of PM

dciodvfy is happy!

fedorov committed:
BUG: fix initialization of the ITK PM output geometry

rows/columns were flipped on initialization

fedorov committed:
ENH: added reminder comment to revisit modality initialization

fedorov committed:
BUG: updates to account for standard updates

 - PixelContrast is no longer needed for PM
 - use standard code for fitting model type

fedorov committed:
BUG: initialize origin from ITK image

fedorov committed:
BUG: fix initialization of ReferencedSeriesSequence

fedorov committed:
BUG: pass the initialized derivation FG to addFrame()

fedorov committed:
ENH: initialize DerivationCode from input JSON

fedorov committed:
BUG: initialize series-specific attributes from JSON

Round-trip tests are now passing

fedorov committed:
BUG: fix initialization of DerivationImageSequence

If DerivationImageSequence is not empty for at least one frame, it must be present
for each of the frames.

fedorov committed:
touch to trigger circleci build

fedorov committed:
ENH: add codecov badge

che85 committed:
ENH: fixed extent computation for both paramap and seg (issue #275)

* initializeVolumeGeometryFromDICOM now takes functional group instead
  of an iodImage. No need to be defined as a template method anymore
* extracted methods

che85 committed:
ENH: reusing initializeVolumeGeometry and reorganizing

* renamed method createParametricMap to createDICOMParametricMap and
  added createITKParametricMap

fedorov commented on libsrc/ParametricMapObject.cpp:
```diff
@@ -140,6 +140,48 @@ int ParametricMapObject::createParametricMap() {
+  DPMParametricMapIOD::Frames<Float32PixelType> frames = *OFget<DPMParametricMapIOD::Frames<Float32PixelType> >(&obj);
+
+  FGInterface &fgInterface = parametricMap->getFunctionalGroups();
+  for(int frameId=0;frameId<volumeGeometry.extent[2];frameId++){
```
would it make sense to have a function that takes FGInterface, and returns a volume-reconstructed ITK volume?

che85 replied:
I am not sure because for DICOM we are doing it the same way ....

fedorov replied:
Sorry, I don't understand!

fedorov replied:
So you decided that it is not worth factoring this function out in the end?

che85 replied:
If we decide to do that it would also need to take frames as a parameter next to the FGInterface.

fedorov commented on libsrc/MultiframeObject.cpp:
```diff
@@ -77,6 +77,60 @@ int MultiframeObject::initializeVolumeGeometryFromITK(DummyImageType::Pointer im
+
+  cout << "computed extent: " << computedVolumeExtent << "/" << spacing[2] << endl;
+
+  extent[2] = round(computedVolumeExtent/spacing[2] + 1);
```
Can you please add a comment above this line, and a reference to https://github.com/QIICR/dcmqi/issues/275?

fedorov commented:
@che85 can you test the updated code for some number of cases from QIN-HEADNECK?

Can you also change from `ceil` to `round` in the current dcmqi code (the one in the current working code base)?

che85 commented:
@fedorov Yeah I will test some cases ..

che85 committed:
ENH: added comment for computation of extent including reference

fedorov committed:
BUG: do not cache codecov build directory

che85 committed:
ENH: separated matching frames with segmentId/slice and writing segment images

fedorov commented:
Thanks! Let me know when it is ready for integration.

che85 committed:
ENH: renamed method, removed some comments and simplified

che85 committed:
ENH: added method for creating CodeSequenceMacro from metadata attribute

che85 commented:
Ready to merge. Shall I start with conversion of itk to dcm segmentations?

che85 committed:
ENH: fixed type and renamed type ShortImageType to ShortITKImageType

* according to the way it's done for parametric maps

che85 committed:
ENH: added missing type change

fedorov commented on libsrc/SegmentationImageObject.cpp:
```diff
@@ -61,70 +62,78 @@ int SegmentationImageObject::iterateOverFramesAndMatchSlices() {
+      unpackedFrame = new DcmIODTypes::Frame(*frame);
+    }
+
+    for (unsigned row = 0; row < volumeGeometry.extent[1]; row++) {
```
Does it make sense to avoid duplication of this loop with the one here: https://github.com/fedorov/dcmqi/pull/5/files#diff-0d22ece1e114de7c6c723ba057d07f76R157?

che85 replied:
I am not sure about that. To be those loops are kind of similar but they are using different pixel types, and also the frame type is not the same (DcmIODTypes::Frame *unpackedFrame and Float32PixelType *frame). Therefore the pixel accessor is different as well. Maybe I am missing something...

che85 committed:
ENH: factored out creation of itkimage from frames

fedorov committed:
ENH: switch to 3-clause BSD license

See discussion in #274.

fedorov committed:
ENH: update license wording in README

fedorov committed:
STYLE: fix typo

fedorov committed:
ENH: further simplify mentioning of the license

fedorov committed:
WIP: adding attribute to link segmentAttributes sub-arrays to specific files

Compiles, need to test

Aims to address #253

fedorov committed:
ENH: make mkdir happy if the directory exists

fedorov committed:
BUG: reduce intensity tolerance test parameter

Default of 2 led to erroneous results of SEG conversion testing

Fixing this bug exposed a test error in the ITK-SEG roundtrip process:
 - input ITK images all have labels 1
 - ITK images produced from SEG have labels that match segment number in SEG
 - there are exactly 3 segments, and since the tolerance was set to 2, the
   non-zero difference between the ITK baselines and the images produced from SEG
   were not detected by the test

The issue above was fixed by updating the baseline images to have the pixel value
matching the segment number, and updating the JSON file used by the test to point to
the updated labels.

fedorov committed:
ENH: finalize file order mapping

Added missing functionality, added testing

fedorov committed:
BUG: add segmentAttributesFileMapping

Remove the attribute added initially. As before, the question is why our schema
validation didn't handle the issue during testing?!..

fedorov committed:
BUG: fix duplicate test name

che85 committed:
ENH: updated dcmtk version

che85 committed:
ENH: started initialization from itk to DICOM for segmentations

- moved initialization of EquipmentInfo to specific class since it's
  different for PM and SEG
- auto resolving headers and source files in CMakeLists.txt

che85 committed:
ENH: proceeding with further extraction and generalizing code

* removed additional parameter ImageVolumeGeometry since it uses the
  volumeGeometry member of the class
* extracted duplicated code into one method
* extracted boolean method for checking if derivation FG is required
* templated getSliceMapForSegmentation2DerivationImage over imagetype
* extracted boolean method for checking if hasDerivationImages

che85 committed:
Merge branch 'frame-handling-refactoring' into rebased-frame-handling-refactoring

che85 commented:
@fedorov pulled after merging, compiled successfully and tests are passing

fedorov commented:
Great, thanks!

fedorov commented:
I am SO confused by the commit history here, no idea how this works (or is supposed to work). I think I am going to just merge and then see what is going on.

fedorov commented:
It's a total mess - now we have duplicate commits! Not sure what is the best way to clean up this mess.
